File: server/src/main/java/org/dan/jadalnia/app/order/OrderAggregator.java
package org.dan.jadalnia.app.order;

import com.google.common.util.concurrent.AtomicLongMap;
import lombok.val;
import org.dan.jadalnia.app.festival.menu.DishName;
import org.dan.jadalnia.app.order.pojo.PaidOrder;

import java.util.Collection;

public class OrderAggregator {
    public AtomicLongMap<DishName> aggregate(Collection<PaidOrder> paidOrders) {
        val counters = AtomicLongMap.<DishName>create();
        paidOrders.forEach(order -> {
            order.getItems().forEach(item -> {
                counters.addAndGet(item.getName(), item.getQuantity());
                item.getAdditions().forEach(addition ->
                        counters.addAndGet(addition.getName(), addition.getQuantity()));
            });
        });
        return counters;
    }
}

File: service/src/main/kotlin/org/dan/jadalnia/app/order/OrderAggregator.kt
package org.dan.jadalnia.app.order;

import com.google.common.util.concurrent.AtomicLongMap
import org.dan.jadalnia.app.festival.menu.DishName
import org.dan.jadalnia.app.order.pojo.PaidOrder

class OrderAggregator {
    fun aggregate(paidOrders: Collection<PaidOrder>): AtomicLongMap<DishName> {
        val counters = AtomicLongMap.create<DishName>()
        paidOrders.forEach({ order ->
            order.items.forEach({ item ->
                counters.addAndGet(item.name, item.quantity.toLong())
                item.additions.forEach({ addition ->
                    counters.addAndGet(addition.name, addition.quantity.toLong())
                })
            })
        })
        return counters;
    }
}
